pkg/scheduler/scheduler.go

```diff
@@ -780,11 +783,23 @@ type podConditionUpdaterImpl struct {
 
 func (p *podConditionUpdaterImpl) update(pod *v1.Pod, condition *v1.PodCondition) error {
 	klog.V(3).Infof("Updating pod condition for %s/%s to (%s==%s, Reason=%s)", pod.Namespace, pod.Name, condition.Type, condition.Status, condition.Reason)
-	if podutil.UpdatePodCondition(&pod.Status, condition) {
-		_, err := p.Client.CoreV1().Pods(pod.Namespace).UpdateStatus(context.TODO(), pod, metav1.UpdateOptions{})
+	oldData, err := json.Marshal(pod)
+	if err != nil {
 		return err
 	}
-	return nil
+	if !podutil.UpdatePodCondition(&pod.Status, condition) {
+		return nil
+	}
+	newData, err := json.Marshal(pod)
+	if err != nil {
+		return err
+	}
+	patchBytes, err := strategicpatch.CreateTwoWayMergePatch(oldData, newData, &v1.Pod{})
```

Comment:
any idea how expensive it is to marshal two objects and create a patch compared to making a PUT on the status directly?

Reply:
Can you share a sample of this output? I wonder if we can Marshal smaller portions (like directly Status). Also, we should not use json.Marshal, but get an Encoder from apimachinery packages.

Reply:
Sure, I added some logging and ran the tests to capture the patch content...

Setting the nominated node:
```
{"status":{"nominatedNodeName":"node1"}}
```

Clearing the nominated node:
```
{"status":{"nominatedNodeName":null}}
```

Updating the pod conditions:
```
{"status":{"conditions":[{"lastProbeTime":null,"lastTransitionTime":"2020-05-12T16:18:14Z","message":"0/1 nodes are available: 1 Insufficient cpu, 1 Insufficient memory.","reason":"Unschedulable","status":"False","type":"PodScheduled"}]}}
```

---

> Also, we should not use json.Marshal, but get an Encoder from apimachinery packages.

Almost all of the places I looked where `CreateTwoWayMergePatch` is being used seem to be using `json.Marshal` which is why I used it. If you think it is better to use the Encoder from apimachinery, I can change it though.

Reply:
I think I see what you're saying though about marshaling smaller portions, can we just do something like this...
```
	oldData, err := json.Marshal(podCopy.Status)
	if err != nil {
		return err
	}
	podCopy.Status.NominatedNodeName = nominatedNodeName
	newData, err := json.Marshal(podCopy.Status)
	if err != nil {
		return err
	}
	patchBytes, err := strategicpatch.CreateTwoWayMergePatch(oldData, newData, &v1.PodStatus{})
```

Let me give it a try...

Reply:
The other thing is we know exactly what is changing, so do we even need to marshal at all?  Can we just construct `{"status":{"nominatedNodeName":"node1"}}` directly?

Reply:
Feels like a premature optimization to me, but I don't know how sensitive scheduler is to performance

Reply:
I was reading the documentation of `CreateTwoWayMergePatch`, which is supposed to accept Precondition functions. Are we sure this fails if the initial `nominatedNodeName` doesn't match the original we have?

Reply:
The preconditions are optional and most callers don't use it.  The only one I found that used it was in [editoptions](https://github.com/kubernetes/kubernetes/blob/50c8f73a4b23b18ec0f87a45748a063c740055c4/staging/src/k8s.io/kubectl/pkg/cmd/util/editor/editoptions.go#L617).

You're right though, we should make sure the behavior is exactly what is intended. It is using a "strategic merge" which I believe does what is described here:
https://kubernetes.io/docs/tasks/run-application/update-api-object-kubectl-patch/#notes-on-the-strategic-merge-patch

I'll look into this a little more to make sure this will update things as expected, maybe write an integration test around that, or update an existing integration test

Reply:
FWIW, the Conditions metadata says this uses the "merge" type, keying on `type`... so I think that means, replacing any existing condition of the same type, but leaving all other existing conditions alone, but I can check to make sure.

```
Conditions []PodCondition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,2,rep,name=conditions"`
```

Reply:
Oh.... I see. What about `nominatedNodeName`?

Reply:
nominatedNodeName doesn't need a patch strategy, since it is not an array... it just replaces it.